pkg/services/secrets/kvstore/plugin.go

```diff
@@ -22,43 +23,67 @@ var (
 	errPluginNotInstalled     = errors.New("remote secret managements plugin disabled because there is no installed plugin of type `secretsmanager`")
 )
 
-// secretsKVStorePlugin provides a key/value store backed by the Grafana plugin gRPC interface
-type secretsKVStorePlugin struct {
+// SecretsKVStorePlugin provides a key/value store backed by the Grafana plugin gRPC interface
+type SecretsKVStorePlugin struct {
 	log                            log.Logger
 	secretsPlugin                  smp.SecretsManagerPlugin
 	secretsService                 secrets.Service
 	kvstore                        *kvstore.NamespacedKVStore
 	backwardsCompatibilityDisabled bool
-	fallback                       SecretsKVStore
+	fallbackEnabled                bool
+	fallbackStore                  SecretsKVStore
+}
+
+func NewPluginSecretsKVStore(
+	secretsPlugin smp.SecretsManagerPlugin,
+	secretsService secrets.Service,
+	kvstore *kvstore.NamespacedKVStore,
+	features featuremgmt.FeatureToggles,
+	fallback SecretsKVStore,
+	logger log.Logger,
+) *SecretsKVStorePlugin {
+	return &SecretsKVStorePlugin{
+		secretsPlugin:                  secretsPlugin,
+		secretsService:                 secretsService,
+		log:                            logger,
+		kvstore:                        kvstore,
+		backwardsCompatibilityDisabled: features.IsEnabled(featuremgmt.FlagDisableSecretsCompatibility),
+		fallbackStore:                  fallback,
+	}
 }
 
 // Get an item from the store
 // If it is the first time a secret has been retrieved and backwards compatibility is disabled, mark plugin startup errors fatal
-func (kv *secretsKVStorePlugin) Get(ctx context.Context, orgId int64, namespace string, typ string) (string, bool, error) {
+func (kv *SecretsKVStorePlugin) Get(ctx context.Context, orgId int64, namespace string, typ string) (string, bool, error) {
 	req := &smp.GetSecretRequest{
 		KeyDescriptor: &smp.Key{
 			OrgId:     orgId,
 			Namespace: namespace,
 			Type:      typ,
 		},
 	}
+
 	res, err := kv.secretsPlugin.GetSecret(ctx, req)
-	if err != nil {
-		return "", false, err
-	} else if res.UserFriendlyError != "" {
+	if res.UserFriendlyError != "" {
 		err = wrapUserFriendlySecretError(res.UserFriendlyError)
 	}
 
 	if res.Exists {
 		updateFatalFlag(ctx, *kv)
 	}
 
+	if kv.fallbackEnabled {
```

Comment:
Good idea, I was thinking that we should just use the fallback store, but it makes sense to try to get it from the primary store first and use this as a backup 👍